File: golang/peeking_iterator.go
package golang

/*   Below is the interface for Iterator, which is already defined for you.
 *
 *   type Iterator struct {
 *
 *   }
 *
 *   func (this *Iterator) hasNext() bool {
 *		// Returns true if the iteration has more elements.
 *   }
 *
 *   func (this *Iterator) next() int {
 *		// Returns the next element in the iteration.
 *   }
 */

// PeekingIterator is struct for PeekingIterator class
type PeekingIterator struct {
    Iter *Iterator
    Next int
    HasPeeked bool
}

// PeekingIteratorConstructor object constructor
func PeekingIteratorConstructor(iter *Iterator) *PeekingIterator {
    obj := &PeekingIterator{
        Iter: iter,
        HasPeeked: false,
    }
    
    return obj
}

func (obj *PeekingIterator) hasNext() bool {
    return obj.HasPeeked || obj.Iter.hasNext()
}

func (obj *PeekingIterator) next() int {
    if !obj.HasPeeked {
        obj.Next = obj.Iter.next()
    }
    obj.HasPeeked = false
    
    return obj.Next
}

func (obj *PeekingIterator) peek() int {
    if !obj.HasPeeked {
        obj.Next = obj.Iter.next()
    }
    obj.HasPeeked = true
    
    return obj.Next
}

File: python3/peeking_iterator.py
# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator:
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """

class PeekingIterator:
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.iter = iterator
        self.hasPeeked = False

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if not self.hasPeeked:
            self.n = self.iter.next()
        self.hasPeeked = True

        return self.n

    def next(self):
        """
        :rtype: int
        """
        if not self.hasPeeked:
            self.n = self.iter.next()
        self.hasPeeked = False

        return self.n

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.hasPeeked or self.iter.hasNext()


# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].
